File: project/backend/app/models/provider.py
"""
Provider model for healthcare providers (may or may not have system accounts)
"""
from datetime import datetime

from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from app.db.base import Base


class Provider(Base):
    """
    A healthcare provider (physician, therapist, etc.).

    user_id is nullable because providers imported from external datasets may
    not have system accounts.  encounter_count and procedure_count are
    pre-aggregated analytics fields from the import pipeline.
    """
    __tablename__ = "providers"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True, index=True,
    )
    organization_id = Column(
        Integer, ForeignKey("organizations.id", ondelete="SET NULL"),
        nullable=True, index=True,
    )
    name = Column(String(255), nullable=False)
    gender = Column(String(10), nullable=True)
    speciality = Column(String(100), nullable=True, index=True)
    address = Column(Text, nullable=True)
    city = Column(String(100), nullable=True, index=True)
    state = Column(String(50), nullable=True)
    zip = Column(String(20), nullable=True)
    lat = Column(Float, nullable=True)
    lon = Column(Float, nullable=True)
    encounter_count = Column(Integer, nullable=False, default=0, server_default="0")
    procedure_count = Column(Integer, nullable=False, default=0, server_default="0")
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    # Relationships
    user = relationship("User", foreign_keys=[user_id])
    organization = relationship("Organization", back_populates="providers")

    def __repr__(self) -> str:
        return f"<Provider(id={self.id}, name='{self.name}', speciality='{self.speciality}')>"

File: project/mobile/lib/models/provider.dart
/// Provider model representing a healthcare provider.
class Provider {
  final int id;
  final int? userId;
  final int? organizationId;
  final String? organizationName;
  final String name;
  final String? gender;
  final String? speciality;
  final String? address;
  final String? city;
  final String? state;
  final String? zip;
  final double? lat;
  final double? lon;
  final int encounterCount;
  final int procedureCount;
  final DateTime createdAt;

  Provider({
    required this.id,
    this.userId,
    this.organizationId,
    this.organizationName,
    required this.name,
    this.gender,
    this.speciality,
    this.address,
    this.city,
    this.state,
    this.zip,
    this.lat,
    this.lon,
    required this.encounterCount,
    required this.procedureCount,
    required this.createdAt,
  });

  factory Provider.fromJson(Map<String, dynamic> json) {
    return Provider(
      id: json['id'] as int,
      userId: json['user_id'] as int?,
      organizationId: json['organization_id'] as int?,
      organizationName: json['organization_name'] as String?,
      name: json['name'] as String,
      gender: json['gender'] as String?,
      speciality: json['speciality'] as String?,
      address: json['address'] as String?,
      city: json['city'] as String?,
      state: json['state'] as String?,
      zip: json['zip'] as String?,
      lat: (json['lat'] as num?)?.toDouble(),
      lon: (json['lon'] as num?)?.toDouble(),
      encounterCount: json['encounter_count'] as int? ?? 0,
      procedureCount: json['procedure_count'] as int? ?? 0,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  /// Formatted city/state string for list display.
  String get location {
    if (city != null && state != null) return '$city, $state';
    if (city != null) return city!;
    if (state != null) return state!;
    return 'Unknown';
  }
}
